plutus-ir/src/Language/PlutusIR/Parser.hs

```diff
@@ -0,0 +1,344 @@
+
+kind :: Parser (Kind SourcePos)
+kind = inParens (typeKind <|> sizeKind <|> funKind)
+    where
+        typeKind = Type <$> reservedWord "type"
+        sizeKind = Size <$> reservedWord "size"
+        funKind  = KindArrow <$> reservedWord "fun" <*> kind <*> kind
+
+typ :: Parser (Type TyName SourcePos)
+typ = (tyVar >>= (\n -> return $ TyVar (nameAttribute $ unTyName n) n))
+    <|> (inParens $ funType <|> allType <|> lamType <|> ifixType <|> conType)
+    <|> inBrackets appType
+
+varDecl :: Parser (VarDecl TyName Name SourcePos)
+varDecl = inParens $ VarDecl <$> reservedWord "vardecl" <*> var <*> typ
+
+tyVarDecl :: Parser (TyVarDecl TyName SourcePos)
+tyVarDecl = inParens $ TyVarDecl <$> reservedWord "tyvardecl" <*> tyVar <*> kind
+
+datatype :: Parser (Datatype TyName Name SourcePos)
+datatype = inParens $ Datatype <$> reservedWord "datatype"
+    <*> tyVarDecl
+    <*> some tyVarDecl
+    <*> var
+    <*> some varDecl
+
+binding :: Parser (Binding TyName Name SourcePos)
+binding =  inParens $
+    (try $ reservedWord "termbind" >> TermBind <$> getPosition <*> varDecl <*> term)
+    <|> (reservedWord "typebind" >> TypeBind <$> getPosition <*> tyVarDecl <*> typ)
+    <|> (reservedWord "datatypebind" >> DatatypeBind <$> getPosition <*> datatype)
+
+absTerm :: Parser (Term TyName Name SourcePos)
+absTerm = TyAbs <$> reservedWord "abs" <*> tyVar <*> kind <*> term
+
+lamTerm :: Parser (Term TyName Name SourcePos)
+lamTerm = LamAbs <$> reservedWord "lam" <*> name <*> typ <*> term
+
+conTerm :: Parser (Term TyName Name SourcePos)
+conTerm = Constant <$> reservedWord "con" <*> constant
+
+iwrapTerm :: Parser (Term TyName Name SourcePos)
+iwrapTerm = IWrap <$> reservedWord "iwrap" <*> typ <*> typ <*> term
+
+builtinTerm :: Parser (Term TyName Name SourcePos)
+builtinTerm = Builtin <$> reservedWord "builtin" <*> builtinVar
+
+unwrapTerm :: Parser (Term TyName Name SourcePos)
+unwrapTerm = Unwrap <$> reservedWord "unwrap" <*> term
+
+errorTerm :: Parser (Term TyName Name SourcePos)
+errorTerm = Error <$> reservedWord "error" <*> typ
+
+letTerm :: Parser (Term TyName Name SourcePos)
+letTerm = Let <$> reservedWord "let" <*> recursivity <*> some (try binding) <*> term
+
+appTerm :: Parser (Term TyName Name SourcePos)
+appTerm = do
+    pos  <- getPosition
+    fn   <- term
+    args <- some term
+    pure $ foldl' (Apply pos) fn args
+
+tyInstTerm :: Parser (Term TyName Name SourcePos)
+tyInstTerm = do
+    pos  <- getPosition
+    fn   <- term
+    args <- some typ
+    pure $ foldl' (TyInst pos) fn args
+
+term :: Parser (Term TyName Name SourcePos)
+term = (var >>= (\n -> return $ Var (nameAttribute n) n))
+    <|> (inParens $ absTerm <|> lamTerm <|> conTerm <|> iwrapTerm <|> builtinTerm <|> unwrapTerm <|> errorTerm <|> letTerm)
+    <|> inBraces tyInstTerm
+    <|> inBrackets appTerm
+
+-- Note that PIR programs do not actually carry a version number
+-- we (optionally) parse it all the same so we can parse all PLC code
```

Comment:
This problem goes away if you have a separate PLC parser and the PIR parser only uses the *term* part of it.

Reply:
There's no real problem, we could just make the parser reject all PLC programs, but it feels a bit odd to go out of our way to make it not compatible with PLC since all PLC programs are valid PIR programs (minus version numbers).